cli/azd/pkg/messaging/topic.go

```diff
@@ -0,0 +1,59 @@
+package messaging
+
+import (
+	"context"
+
+	"golang.org/x/exp/slices"
+)
+
+type Topic struct {
+	Name        string
+	channel     chan *Message
+	subscribers []*Subscription
+}
+
+func NewTopic(name string) *Topic {
+	return &Topic{
+		Name:        name,
+		channel:     make(chan *Message),
+		subscribers: []*Subscription{},
+	}
+}
+
+func (t *Topic) Subscribe(ctx context.Context, filter MessageFilter, handler MessageHandler) *Subscription {
+	subscription := NewSubscription(t, filter, handler)
+	t.subscribers = append(t.subscribers, subscription)
+
+	return subscription
+}
+
+func (t *Topic) Unsubscribe(subscription *Subscription) {
+	index := slices.IndexFunc(t.subscribers, func(s *Subscription) bool {
+		return s == subscription
+	})
+
+	if index < 0 {
+		return
+	}
+
+	// Remove subscription from t.subscribers
+	t.subscribers = append(t.subscribers[:index], t.subscribers[index+1:]...)
+}
+
+func (t *Topic) Send(ctx context.Context, msg *Message) {
+	t.channel <- msg
+}
+
+func (t *Topic) Close(ctx context.Context) {
+	close(t.channel)
+}
+
+func (t *Topic) init(ctx context.Context) {
+	go func() {
+		for msg := range t.channel {
+			for _, subscriber := range t.subscribers {
+				subscriber.receive(msg)
+			}
+		}
+	}()
```

Comment:
After a topic is initialized it will start reading messages off the channel and notify any topic subscriptions.